src/rust/engine/src/externs/interface.rs

```diff
@@ -566,6 +578,90 @@ py_class!(class PyNailgunServer |py| {
     }
 });
 
+py_class!(class PyNailgunClient |py| {
+  data executor: PyExecutor;
+  data port: u16;
+
+  def execute(&self, py_signal_fn: PyObject, command: PyString, args: PyList, env: PyDict) -> CPyResult<PyInt> {
+
+    let command: String = command.to_string(py)
+      .map(|s| s.into())?;
+
+    let args: Vec<String> = args
+      .iter(py)
+      .map(|arg: PyObject| arg.cast_into::<PyString>(py)
+        .map_err(|err| err.into())
+        .and_then(|s: PyString| s.to_string(py).map(|rust_str| rust_str.into_owned()))
+      )
+      .collect::<Result<Vec<_>, _>>()?;
+
+    let env_list: Vec<(String, String)> = env
+    .items(py)
+    .into_iter()
+    .map(|(k, v): (PyObject, PyObject)| -> Result<(String, String), PyErr> {
+      let k: String = k.extract::<String>(py)?;
+      let v: String = v.extract::<String>(py)?;
+      Ok((k, v))
+    })
+    .collect::<Result<Vec<_>, _>>()?;
+
+    let port = *self.port(py);
+    let executor_ptr = self.executor(py);
+    let python_signal_fn: Value = py_signal_fn.into();
+
+    let (exit_sender, exit_receiver) = oneshot::channel();
+    let nailgun_fut = nailgun::client_execute(
+      port,
+      command,
+      args,
+      env_list,
+      exit_receiver,
+    );
+
+    let exit_code: Result<i32, String> = with_executor(py, executor_ptr, |executor| {
+        let (sender, receiver) = mpsc::channel();
+
+        let _spawned_fut = executor.spawn(async move {
+          let exit_code = nailgun_fut.await;
+          let _ = sender.send(exit_code);
+        });
```

Comment:
You should not need to spawn a task. If you got an error like "not running on the `tokio` runtime", you should "enter" the runtime (which sets some thread-local state that tokio uses to find the runtime):
```
let result = executor.enter(|| {
  // ... do some stuff on the runtime 
})
```

But also, using `executor.block_on` should enter the runtime... so I think that you might be able to replace most of this function body with:
```
let exit_code = executor.block_on(nailgun_fut);
```